### PR TITLE
Fixed #31046 -- Allowed RelatedManager.add()/create()/set() to accept callable values in through_defaults.

### PR DESCRIPTION
**Pull request by bmispelon**
ticket-31046

### CODE REVIEW

File: django/db/models/fields/related_descriptors.py [hunk 1]
```diff
@@ -1113,7 +1114,7 @@ def _add_items(self, source_field_name, target_field_name, *objs, through_defaul
             # source_field_name: the PK fieldname in join table for the source object
             # target_field_name: the PK fieldname in join table for the target object
             # *objs - objects to add. Either object instances, or primary keys of object instances.
-            through_defaults = through_defaults or {}
+            through_defaults = dict(resolve_callables(through_defaults or {}))
```
Comment (charettes) [inline django/db/models/fields/related_descriptors.py]: I was about to suggest to move this line within the `if objs` branch but doing an early return on `if not objs` like we do in `_remove_items` is likely even better. That should likely be done in another commit.

File: docs/ref/models/relations.txt [hunk 2]
```diff
@@ -73,6 +73,11 @@ Related objects reference
         :ref:`intermediate model <intermediary-manytomany>` instance(s), if
         needed.
 
+        .. versionchanged:: 3.1
+
+            You can use callables as values in the ``through_defaults`` dictionary
+            and they will be evaluated just before adding the model.
```
Comment (charettes) [inline docs/ref/models/relations.txt]: Should we document that they'll be called once and not once per `objs`?

File: django/db/models/fields/related_descriptors.py (hunk 1 above)
Comment (bmispelon) [reply django/db/models/fields/related_descriptors.py]: That's a good idea, thanks.

I'll do that change in a separate commit and rearrange the history so that it comes first.

File: docs/ref/models/relations.txt (hunk 2 above)
Comment (bmispelon) [reply docs/ref/models/relations.txt]: Good point, I hadn't thought about that. I'll rephrase those paragraphs to make it more obvious what "evaluated" means.

Comment (bmispelon) [reply docs/ref/models/relations.txt]: I've (force)pushed an update to try and clarify the situation.

**Comment by felixxm** [comment]
@bmispelon Thanks for this patch :+1: I fixed `flake8` errors, rebased from master, added test for `create()`, reorganized commits, and pushed edits to tests and docs.

**Comment by bmispelon** [comment]
Looking good, thanks! :sparkles:
